sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,194 @@
+	wd:           nil,
+	pathList:     make(map[string]int32),
+	addWatchChan: make(chan *groupToWatch),
+}
+
+func (iw *inotifyWatcher) addWatch(newWatch *groupToWatch) {
+	var folderList []string
+	for _, fullPath := range newWatch.folders {
+		// It's not possible to use inotify.InDeleteSelf in /sys/fs because it
+		// isn't triggered, so we must monitor the parent folder and use InDelete
+		basePath := path.Dir(fullPath)
+		if _, exists := iw.pathList[basePath]; !exists {
+			iw.pathList[basePath] = 0
+			if err := iw.wd.AddWatch(basePath, inotify.InDelete); err != nil {
+				delete(iw.pathList, basePath)
+				continue
+			}
+		}
+		iw.pathList[fullPath]++
+		if osutil.FileExists(fullPath) {
+			folderList = append(folderList, fullPath)
+		} else {
+			iw.removePath(fullPath)
+		}
+	}
+	if len(folderList) == 0 {
+		newWatch.channel <- newWatch.name
+	} else {
+		newWatch.folders = folderList
+		iw.groupList = append(iw.groupList, newWatch)
+	}
+}
+
+func (iw *inotifyWatcher) removePath(fullPath string) {
+	iw.pathList[fullPath]--
+	if iw.pathList[fullPath] == 0 {
+		iw.wd.RemoveWatch(fullPath)
+		delete(iw.pathList, fullPath)
+	}
+}
+
+func (iw *inotifyWatcher) processEvent(watch *groupToWatch, event *inotify.Event) bool {
+	var tmpFolders []string
+	// every time an "InDelete" event is received, check if this CGroup is waiting for
+	// that file/folder deletion event
+	for _, fullPath := range watch.folders {
+		if fullPath != event.Name {
+			tmpFolders = append(tmpFolders, fullPath)
+		} else {
+			// if the folder name is in the list of folders to monitor, decrement the
+			// parent's usage counter, and remove it from the list of folders to watch
+			// in this CGroup (by not adding it to tmpFolders)
+			iw.removePath(fullPath)
+		}
+	}
+	watch.folders = tmpFolders
+	if len(tmpFolders) == 0 {
+		// if all the files/folders of this CGroup have been deleted, notify the
+		// client that it is done.
+		watch.channel <- watch.name
+		// By returning FALSE, the caller is notified to remove this CGroup from
+		// the list of watched CGroups.
+		return false
+	}
+	// Returning TRUE means that this CGroup still have files/folders to be
+	// monitored, and must be kept in the CGroup list.
+	return true
+}
+
+func (iw *inotifyWatcher) watcherMainLoop() {
+	for {
+		select {
+		case event := <-iw.wd.Event:
+			if event.Mask&inotify.InDelete == 0 {
+				continue
+			}
+			var newGroupList []*groupToWatch
+			for _, watch := range iw.groupList {
+				if iw.processEvent(watch, event) {
+					newGroupList = append(newGroupList, watch)
+				}
+			}
+			iw.groupList = newGroupList
+		case newWatch := <-iw.addWatchChan:
+			iw.addWatch(newWatch)
+		}
+	}
+}
+
+// MonitorFullDelete allows to monitor a group of files/folders
+// and, when all of them have been deleted, emits the specified name through the channel.
+func (iw *inotifyWatcher) monitorFullDelete(name string, folders []string, channel chan string) error {
+	iw.doOnce.Do(func() {
```

Comment:
I'm not sure why we're choosing to implement this as a thread-safe singleton here but it'll make testing it more complicated and hacky. I haven't done a deep review of the tests (I will when I re-review) but I see some `time.Sleep()`s which mean that the tests will be prone to race conditions (or slow, if the waiting times are conservative). Is there a reason to do this here? I'm assuming it's related to how it's meant to be used later but I'm lacking that context.

Reply:
The singleton is because each opened inotify instance consumes one file descriptor. Originally I opened one instance for each test, but was asked to change it into having a single inotify instance. I think that it was in a previous iteration of the idea, in other MR.